contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,400 @@
+    uint256 public immutable minStakingDeposit;
+    // Liveness ratio in the format of 1e18
+    uint256 public immutable livenessRatio;
+    // ServiceRegistry contract address
+    address public immutable serviceRegistry;
+
+    // Token / ETH balance
+    uint256 public balance;
+    // Token / ETH available rewards
+    uint256 public availableRewards;
+    // Timestamp of the last checkpoint
+    uint256 public tsCheckpoint;
+    // Mapping of serviceId => staking service info
+    mapping (uint256 => ServiceInfo) public mapServiceInfo;
+    // Set of currently staking serviceIds
+    uint256[] public setServiceIds;
+
+    /// @dev ServiceStakingBase constructor.
+    /// @param _maxNumServices Maximum number of staking services.
+    /// @param _rewardsPerSecond Staking rewards per second (in single digits).
+    /// @param _minStakingDeposit Minimum staking deposit for a service to be eligible to stake.
+    /// @param _livenessRatio Liveness ratio: number of nonces per second (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    constructor(
+        uint256 _maxNumServices,
+        uint256 _rewardsPerSecond,
+        uint256 _minStakingDeposit,
+        uint256 _livenessRatio,
+        address _serviceRegistry)
+    {
+        // Initial checks
+        if (_maxNumServices == 0 || _rewardsPerSecond == 0 || _minStakingDeposit == 0 || _livenessRatio == 0) {
+            revert ZeroValue();
+        }
+        if (_serviceRegistry == address(0)) {
+            revert ZeroAddress();
+        }
+
+        rewardsPerSecond = _rewardsPerSecond;
+        minStakingDeposit = _minStakingDeposit;
+        livenessRatio = _livenessRatio;
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev Checks token / ETH staking deposit.
+    /// @param stakingDeposit Staking deposit.
+    function _checkTokenStakingDeposit(uint256, uint256 stakingDeposit) internal view virtual {
+        // The staking deposit derived from a security deposit value must be greater or equal to the minimum defined one
+        if (stakingDeposit < minStakingDeposit) {
+            revert LowerThan(stakingDeposit, minStakingDeposit);
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal virtual {}
+
+    /// @dev Stakes the service.
+    /// @param serviceId Service Id.
+    function stake(uint256 serviceId) external {
+        // Check if there available rewards
+        if (availableRewards == 0) {
+            revert NoRewardsAvailable();
+        }
+
+        // Check for the maximum number of staking services
+        uint256 numStakingServices = setServiceIds.length;
+        if (numStakingServices == maxNumServices) {
+            revert MaxNumServicesReached(maxNumServices);
+        }
+
+        // Check the service conditions for staking
```

Comment:
I notice a nice optional feature which I'd like to support from the beginning: a staking contract can optionally specify an exact agentID and hash for the service that it requires to be staked. Then here, we check the agentId and service hash as well if and only if the hash is provided in the staking contract.

This would allow staking contracts to express exactly what service they want at which version.

This also raises the question if we should allow the staking contract to specify the number of agent instances and threshold. 

These four things together tie down the exact configuration.

Reply:
ok that's nice